### PR TITLE
New Multiple Locations support in Bulk user upload

### PR DESCRIPTION
**Pull request by sravfeyn**
https://trello.com/c/PI1FHwOi/37-multiple-location-assignment-bulk-upload

@snopoke 
cc @esoergel 

Ready for review. I will need to resolve couple of spec related questions with Sheel before having this merged

### CODE REVIEW

**Comment by czue** [comment]
nice! derek will be pleased

**Review by snopoke — COMMENTED** [review]
Didn't review very closely.

File: corehq/apps/users/bulkupload.py [hunk 1]
```diff
@@ -650,8 +654,6 @@ def _set_progress(progress):
             ) % (_error_message, e.errors))
             ret['errors'].append(_error_message)
 
-    if Domain.get_by_name(domain).supports_multiple_locations_per_user:
-        create_or_update_locations(domain, location_specs, log=ret)
```
Comment (esoergel) [inline corehq/apps/users/bulkupload.py]: Yikes, I didn't even know we did this.  That function should be able to be removed now.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/users/bulkupload.py [hunk 2]
```diff
@@ -476,7 +380,13 @@ def _set_progress(progress):
             uncategorized_data = row.get('uncategorized_data')
             user_id = row.get('user_id')
             username = row.get('username')
-            location_code = row.get('location_code', '')
+            location_codes = row.get('location_code') or []
+            if location_codes and not isinstance(location_codes, list):
```
Comment (sravfeyn) [inline corehq/apps/users/bulkupload.py]: Yeah, this is what I want to do, wanted to get confirmation from Sheel

File: corehq/apps/users/models.py [hunk 3]
```diff
@@ -1777,6 +1777,7 @@ def unset_location_by_id(self, location_id, fall_back_to_next=False):
     def reset_locations(self, location_ids):
         """
         Reset user's assigned_locations to given location_ids and update user data.
+            This should be called after updating primary location via set_location/unset_location
```
Comment (sravfeyn) [inline corehq/apps/users/models.py]: It's hard to enforce it, as it's hard to check from the data whether primary location is updated or not. I was considering to just combining resetting locations and setting primary location into just this one method, but there are multiple ways of checking/enforcing consistency on whether a primary location should be set or not. Instead I thought it's better to leave that logic to callers.

File: corehq/apps/users/bulkupload.py (hunk 2 above)
Comment (sravfeyn) [inline corehq/apps/users/bulkupload.py]: Updated

**Comment by sravfeyn** [comment]
@sheelio 
I have completed most of coding for new [bulk location assignment](https://trello.com/c/PI1FHwOi/37-multiple-location-assignment-bulk-upload). Couple of spec related questions, before I can release this
1. I have removed extra sheet for locations assignment. Same for download
2. location_code column can be just `location_code` one column or multiple columns `location_code 1`, `location_code 2`...
3. Locations won't be created from just bulk user upload. 
4. Is there any restriction on which webusers can assign which locations to which mobile workers? (currently no restrictions are considered)

**Comment by czue** [comment]
not sure if this is time sensitive, but jumping in with quick thoughts on a few of these since @sheelio is offline. let me know if we should be trying to get consensus on this before January

2. seems fine.
3. seems definitely right/desired behavior.
4. this seems tricky and possibly out of scope for current effort if you're using location restrictions. my sense is that the most "correct" behavior would be that you should only be able to modify any user who is assigned to at least one location you can see, and you should only be able to add/remove locations that you have access to. in practice that would be super confusing for a web user so not sure how to figure that out.

**Comment by czue** [comment]
how did we land on the one-column-per-location/group format, anyways? i find it pretty odd (mostly coming from a relational db background), but maybe it's standard practice? (maybe @sheelio the only one who knows that)

**Comment by sravfeyn** [comment]
Re. #4, it's pretty easy to add restrictions in bulk user restrictions. But the spec itself needs some iteration (for e.g., should a user have at least one location assigned manually to be able to use bulk? Can webuser de-assign a location that he/she doesn't have access to). Right now, I can add that a webuser can assign/reassign only those locations that the user has access to. Or we can proceed by adding require-project-admin or no restrictions on webuser at al.

**Comment by czue** [comment]
i think the decision as to what to do right now depends on the time-sensitivity of getting this in. turning off this feature if you have location restrictions enabled seems like the easy short term solution if we need to block on more product capacity to figure out the details though i don't know that impacts current projects.

**Comment by sravfeyn** [comment]
I think Derek is the one who needed this. I will check with him to find the urgency

**Comment by sravfeyn** [comment]
For now I have added logic to disable the feature if location-restritions is enabled. So I am removing the hold-off label

**Comment by esoergel** [comment]
Flagging that the `UploadCommCareUsers` view is not marked as location-safe, so it's not accessible to restricted users anyways.  I did see that you blocked location assignment to the feature flag that was the old means of restricting web-users by location, which is deprecated, and we're working to remove.  I don't think it matters too much how we handle that case, so it's good to be conservative, and I definitely don't think we should be adding any new features there.

At some point (soon?) we'll want to make bulk-user upload location-safe, but that would also require restricting the users you can modify and requiring you to assign locations for new users.  I agree with your assessment that there are some corner cases we should iron out in spec, and that that is out of scope for this PR (no sense in blocking this useful feature for people who can use it now!).

**Comment by dlanedimagi** [comment]
Yes, Derek is very excited about this! and happy to answer any user/pain point/field implementation questions.

**Pull request merged**